rfcs/2022-10-12-14742-flexible-metric-tags.md

```diff
@@ -0,0 +1,306 @@
+
+- [Support duplicate tag and bare tags on metrics](https://github.com/vectordotdev/vector/issues/14742)
+- [DataDog Source -> Filter Transform -> DataDog Sink dropping kube_service tag](https://github.com/vectordotdev/vector/issues/14707)
+- [Invalid custom_tags metrics from sinks datadog_metric](https://github.com/vectordotdev/vector/issues/14239)
+
+## Scope
+
+### In scope
+
+- Changes to the existing metric tag internal and external representations to support multiple
+  values and bare tags.
+
+### Out of scope
+
+- Changes to any other aspect of metric or event representations.
+
+## Pain
+
+When ingesting data from certain metric sources, the tags provided by that source may contain data
+that Vector's data model cannot currently represent. This is causing problems for prospective users
+that want to add Vector to their observability pipelines but cannot do so because it would cause
+data loss. Those sources include:
+
+- `datadog_agent` natively stores tags as bare strings, with optional values separated from the name
+  by a colon.
+- The Prometheus text encoding used by the `prometheus_scrape` source and `prometheus_exporter` sink
+  can support both repeated tag names and bare tags (without a value).
+- The Prometheus `remote_write` encoding similarly supports repeated tag names.
+- OpenTelemetry attributes support keys having arrays of values, which corresponds to repeated tags.
+
+## Proposal
+
+### User Experience
+
+#### External Representation
+
+The existing Protobuf representation of metric tags has a map of string-to-string pairs of
+tags. A new tags map will be added where the values are lists of optional strings, matching the
+semantics given above.
+
+```protobuf
+message Metric {
+  …
+  map<string, string> tags_v1 = 3;
+  message Values {
+    message Value {
+      optional string value = 1;
+    }
+    repeated Value values = 1;
+  }
+  map<string, Values> tags_v2 = 20;
+  …
+}
+```
+
+Similarly, the native JSON encoding for tags is a simple JSON object containing string-to-string
+pairs. This will be enhanced to also allow the `null` value for bare tags, and arrays of either
+strings or `null` for tags with more than one value.
+
+```json
+{
+  "tags": {
+    "single_value":"value",
+    "bare_tag":null,
+    "multi_valued_tag":["value1","value2"],
+    "complex_tag":["value3",null]
+  }
+}
+```
+
+#### Scripting
+
+##### VRL
+
+VRL exposes the tags as a simple map of string values in `.tags`. This will be changed to exposing
+the values as arrays containing either string values or `null`, with the latter representing a bare
+tag. Assignment of a single value to a tag will overwrite all other values for the tag and trigger a
+deprecation warning. Deleting a tag name will remove all tag values with that name. As the values
+will be arrays, these can be used to add or remove individual values. Any tag consisting of an empty
+list will be removed. Since VRL has a "compile" phase, where type information is determined
```

Comment:
Do we need the distinction between `.tags.foo = []` and `.tags.foo = [null]`?

Reply:
It'll get rather messy if we don't. Consider if a script filters a tag set and the result is an empty array. That should be result in all the tags being deleted rather than creating a bare tag.